### ISSUE TITLE
[Feature][Proxy] Squid on-disk cache should survive reboot

### ISSUE THREAD

**Issue by borisneubert** [opnsense/core#1706]
After reboot, the squid cache manager still shows roughly 80,000 objects as before the reboot but the objects are apparently gone. To reproduce: run the command `squidclient -h theopnsensebox mgr:objects|grep GET` before and after reboot.

The squid cache should survive the reboot.

I looked through `/usr/local/etc/rc.d/squid` but I could not find where this can be (un)configured. Squid's documentation says  that squid by default starts with a clean cache, not an empty cache. The cache dir itself survives the reboot as can be seen from the time stamps of the subdirectories and a core dump lying around.

**Comment by fabianfrz** [comment]
you should check if your squid cache is stored on a ramdisk. You can set another directory in the squid configuration or uncheck the setting which makes it a ramdisk in the system settings (reboot required).

**Comment by borisneubert** [comment]
squid.conf: `cache_dir ufs /var/squid/cache 262144 16 256`

`/var` resides on SSD ("local cache" option is enabled):
```
# df
Filesystem      1K-blocks    Used     Avail Capacity  Mounted on
/dev/gpt/rootfs 479420532 6803100 434263792     2%    /
devfs                   1       1         0   100%    /dev
tmpfs            23492152     120  23492032     0%    /tmp
devfs                   1       1         0   100%    /var/dhcpd/dev
```

**Comment by fichtner** [comment]
@borisneubert hey Boris, the cache dir chan be changed, but we also have squid_var_mfs="/var/squid/cache" support nowadays.

The problem: either we fix the squid cache dir again, or we have to let people use the directory override.

What do you think?

**Comment by borisneubert** [comment]
Uh, what is squid_var_mfs="/var/squid/cache" support? I know that the cache dir can be changed (UI: Web Proxy | Administration | General | Local Cache Settings (advanced) | Cache directory location).

Currently the cache resides on the same filesystem as the root filesystem. My preference had been to have a small root fs (e.g. 32 GB) and a large fs on another partition to host the cache. Whether this fs would have been mounted as /var/squid/cache or /something/else is not of importance for me. 

BTW I did not check whether the observed issue still exists in OPNsense 17.7.10.

**Comment by fichtner** [comment]
squid_var_mfs will make that directory persistent, we needed the same thing for ClamAV signature database persistency. But it does not make much sense if the cache dir can be set to somewhere else. The issue is the same on 17.7.11 / 18.1-BETA.

**Comment by borisneubert** [comment]
mfs= memory file system
MFS is OFF for my /var
Having /var (and /var/squid/cache) on MFS makes sense for small systems with SD card or USB sticks as storage medium to avoid early wear off.

I'd thus like to suggest the following:
- /var/squid/cache can be moved anywhere (currently already possible)
- full help for Web Proxy | Administration | General | Local Cache Settings (advanced) | Cache directory location should include a notice that the user should have a look at System | Settings | Miscellaneous | /var RAM disk ("if /var is on RAM disk, cache will emptied on reboot").
- persistence of /var/squid/cache would then be explicitely given if the cache dir is /somewhere/else or if /var is not on RAM disk. 
- the option "Persist cache" would make sense if the cache is explicitely emptied if the option is unset and kept (no matter whether /var is on RAM disk or not) if the option is set.

Still not clear to me why my cache was emptied after reboot although /var is NOT on RAM disk.

**Comment by fichtner** [comment]
Now I am confused. oO

I think points 1-3 are simply documentation issues. And is "Persist cache" an option you propose or am I just not able to find it in the code under that name?

**Comment by borisneubert** [comment]
Points 1-3 are simply documentation issues, right.

I thought that squid_var_mfs="/var/squid/cache" makes the the squid cache dir persist over reboots. My proposal was to 
- either do not have this at all (and stick to the points 1-3) 
- or to expose the option "Persist cache" in the UI, either keeping the cache content over reboots (if checked) or emptying the cache on reboot (if unchecked), no matter where the squid cache dir resides and if /var is on RAM disk or not.

**Comment by fichtner** [comment]
Ok understood. Then we rather leave everything as is.

As to your missing cache, was this during first setup? Does it persist now? I'm asking because there was a bug report recently that could shed light on this: https://github.com/opnsense/core/issues/2008

Basically, if you select local cache, it says "service restart required" and hitting apply does not do that job... It's only picked up after reboot.

**Comment by borisneubert** [comment]
I observed the missing cache in July, most likely during initial tests, but the problem persisted over reboots.

I cannot reproduce the issue today. How did I proceed:
1. Downloaded a large file to desktop. CACHE_MISS
2. Downloaded the file again. CACHE_HIT 
3. Rebooted OPNSense box and waited for Proxy Service to be up and running.
4. Downloaded the file again. CACHE_HIT 

So I would say the issue can be closed.
